_posts/2017-12-27-cheap-docker-swarm-cluster-aws-terraform.md

```diff
@@ -0,0 +1,98 @@
+| Spot | $0.0185 | $13.32 |
+
+Even with full availability, we are still getting cheaper prices than our self-build, but a significant savings when we look at spot pricing.
+
+## Setting it all up
+
+Another major factor we need to discuss is how long it takes to set this stuff up.  We’ve done a simple comparison which makes a compelling case for the cloud.  But where the cloud really shines is in the time savings setting the cluster up.
+
+It look me about 2 hours to build each of my self-hosted boxes, that’s after sourcing all of the parts.  Let’s say a total of 6 hours, assuming everything goes well.
+
+How long does it take on the cloud?  With a little help from [Terraform](https://www.terraform.io/), we can have that same setup in about 5 minutes.
+
+So let’s factor that in as well.  How you value your time is certainly up to you, but to try to keep costs low, let’s assume a $10 minimum hourly wage for the setup time.  That’s an extra $60 of upfront setup time.  Over the course of the lifetime of the servers however, that’s not really that big of a deal, but it does add up.  We could also talk about the cost of maintenance if something goes wrong.  If you’re self hosting, you’re on the hook to fix stuff, unless you need an expensive paper weight.   In the Cloud, in a matter of minutes, you can swap out a bad instance with a fresh new one.
+
+## Show me the code
+
+If you would like to spin something up quickly using spot instances, go ahead and checkout the code.
+
+[https://github.com/rrx/aws-docker-swarm](https://github.com/rrx/aws-docker-swarm)
+
+The code is a fork of some well written terraform by another developer.  It was so well done, that I didn’t think it was worth it to reinvent the wheel.  I just modified it to use spot instances.
+
+Here are some quick instructions to get you up and running.
+
+- Create a keypair in AWS (if you haven’t already). [You can find instructions on how to do this here.](http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-key-pairs.html)
+- Setup your AWS credentials file
+- [Install Terraform.](https://www.terraform.io/intro/getting-started/install.html) Download the appropriate files for your OS and make that file available in your path.
+- ```git clone https://github.com/rrx/aws-docker-swarm```
+- ```cd aws-docker-swarm```
+- Update terraform.tfvars with appropriate values
+- ```make plan```
+- ```make apply```
+- View your instances: ```make swarm-instances```
+- SSH into the swarm master
+- Run benchmarks
+- Take down your cluster. ```make destroy```
+
+## Conclusion
+
+So the bottom line is the cloud is cheap and easy.  And you only pay for what you use.  It’s a new way of thinking about things and it takes some getting used to.  Years ago we didn’t have a choice, we had to buy and provision our servers.  It was expensive to build and maintain.  This new era of cloud computing is cutting those costs down by an order of magnitude and fueling a huge portion of today’s technology innovations.
```

Comment:
Get rid of "So." In fact, start with "Using the cloud for deployment can be cheap and easy, and you only have to pay for what you use." The rest of this is okay.